Import Library


In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from string import ascii_letters
import matplotlib.pyplot as plt
import datetime as dt
import requests
from lxml import html
import math
import json
import re
import os

# 1. County Demographics Data Cleaning and Wrangling


## 1.1 Loading County Demogrphics Dataset

In [2]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("glozab/county-level-us-demographic-data-1990-2020")

print("Path to dataset files:", path)

100%|██████████| 13.1M/13.1M [00:00<00:00, 64.0MB/s]

Extracting files...


Path to dataset files: /root/.cache/kagglehub/datasets/glozab/county-level-us-demographic-data-1990-2020/versions/1


In [3]:
files = os.listdir(path)
print("Files in dataset:", files)

Files in dataset: ['county_demographics.csv']


In [4]:
file_path = os.path.join(path, 'county_demographics.csv')
county_demographics_df = pd.read_csv(file_path)

# print(county_demographics_df.head())
county_demographics_df.describe()


year          fips    population  w_population  b_population  \
count  97287.000000  97287.000000  9.728700e+04  9.728700e+04  9.728700e+04   
mean    2005.009436  30412.859950  9.365957e+04  7.529841e+04  1.252406e+04   
std        8.942825  15147.842995  3.034699e+05  2.259523e+05  5.716518e+04   
min     1990.000000   1001.000000  5.500000e+01  2.900000e+01  0.000000e+00   
25%     1997.000000  18183.000000  1.094800e+04  9.225500e+03  9.500000e+01   
50%     2005.000000  29179.000000  2.490200e+04  2.166100e+04  8.000000e+02   
75%     2013.000000  45083.000000  6.347900e+04  5.624700e+04  5.440000e+03   
max     2020.000000  99999.000000  1.009486e+07  7.301940e+06  1.437609e+06   

        o_population  nh_population  hi_population  na_population  \
count   97287.000000   9.728700e+04   9.728700e+04        97287.0   
mean     1129.919044   8.009754e+04   1.356203e+04            0.0   
std      4815.437070   2.171562e+05   1.058393e+05            0.0   
min         0.000000   4.900000e+01   0.000000e+00            0.0   
25%        40.000000   1.011600e+04   1.710000e+02            0.0   
50%       136.000000   2.327000e+04   6.350000e+02            0.0   
75%       550.000000   5.989000e+04   3.148000e+03            0.0   
max    172454.000000   5.508975e+06   4.868179e+06            0.0   

       male_population  ...  age9_population_ratio  age10_population_ratio  \
count     9.728700e+04  ...           97287.000000            97287.000000   
mean      4.598061e+04  ...               0.068098                0.067555   
std       1.489240e+05  ...               0.011292                0.010417   
min       2.900000e+01  ...               0.008400                0.013420   
25%       5.445000e+03  ...               0.059970                0.060400   
50%       1.231900e+04  ...               0.068510                0.067780   
75%       3.135050e+04  ...               0.075770                0.074500   
max       4.975410e+06  ...               0.160000                0.155960   

       age11_population_ratio  age12_population_ratio  age13_population_ratio  \
count            97287.000000            97287.000000            97287.000000   
mean                 0.064839                0.060595                0.054948   
std                  0.011904                0.014153                0.014746   
min                  0.009150                0.007710                0.005320   
25%                  0.055950                0.048960                0.044160   
50%                  0.065460                0.060580                0.052770   
75%                  0.073120                0.071100                0.064605   
max                  0.153010                0.258060                0.181820   

       age14_population_ratio  age15_population_ratio  age16_population_ratio  \
count            97287.000000            97287.000000            97287.000000   
mean                 0.048050                0.039687                0.030821   
std                  0.013994                0.011871                0.009476   
min                  0.004840                0.002570                0.000000   
25%                  0.038620                0.032010                0.024680   
50%                  0.046150                0.038470                0.029950   
75%                  0.055630                0.045940                0.035970   
max                  0.183910                0.183580                0.143400   

       age17_population_ratio  age18_population_ratio  
count            97287.000000            97287.000000  
mean                 0.021798                0.020099  
std                  0.007503                0.008808  
min                  0.000000                0.000000  
25%                  0.016890                0.014210  
50%                  0.021000                0.018590  
75%                  0.025670                0.024350  
max                  0.144440                0.195400  

[8 rows x 57 

## 1.2 County Demographics Dataset Cleaning

In [6]:
print(county_demographics_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97287 entries, 0 to 97286
Data columns (total 57 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   year                     97287 non-null  int64  
 1   fips                     97287 non-null  int64  
 2   population               97287 non-null  int64  
 3   w_population             97287 non-null  int64  
 4   b_population             97287 non-null  int64  
 5   o_population             97287 non-null  int64  
 6   nh_population            97287 non-null  int64  
 7   hi_population            97287 non-null  int64  
 8   na_population            97287 non-null  int64  
 9   male_population          97287 non-null  int64  
 10  female_population        97287 non-null  int64  
 11  age0_population          97287 non-null  int64  
 12  age1_population          97287 non-null  int64  
 13  age2_population          97287 non-null  int64  
 14  age3_population       

In [21]:
print(county_demographics_df.isna().sum().sort_values(ascending=False))

year                       0
fips                       0
population                 0
w_population               0
b_population               0
o_population               0
nh_population              0
hi_population              0
na_population              0
male_population            0
female_population          0
age0_population            0
age1_population            0
age2_population            0
age3_population            0
age4_population            0
age5_population            0
age6_population            0
age7_population            0
age8_population            0
age9_population            0
age10_population           0
age11_population           0
age12_population           0
age13_population           0
age14_population           0
age15_population           0
age16_population           0
age17_population           0
age18_population           0
w_population_ratio         0
b_population_ratio         0
o_population_ratio         0
nh_population_ratio        0
hi_population_

### Rename Columns for Consistency

In [8]:
county_demographics_df.columns = county_demographics_df.columns.str.strip().str.lower().str.replace(' ', '_')

 ### Handle Duplicates

In [9]:
county_demographics_df = county_demographics_df.drop_duplicates()
county_demographics_df = county_demographics_df.dropna()
county_demographics_df.head()

year  fips  population  w_population  b_population  o_population  \
0  1990  1025       27289         15579         11643            35   
1  1990  1031       40293         32869          6950           160   
2  1990  1041       13598         10068          3516            11   
3  1990  1053       35526         24377         10050          1045   
4  1990  1101      209537        119702         87856           415   

   nh_population  hi_population  na_population  male_population  ...  \
0          27196             93              0            13052  ...   
1          39831            462              0            19673  ...   
2          13576             22              0             6421  ...   
3          35378            148              0            17454  ...   
4         207933           1604              0            98854  ...   

   age9_population_ratio  age10_population_ratio  age11_population_ratio  \
0                0.06263                 0.05552                 0.04998   
1                0.07455                 0.06031                 0.05180   
2                0.06582                 0.05354                 0.04574   
3                0.06820                 0.05694                 0.05174   
4                0.06966                 0.05219                 0.04392   

   age12_population_ratio  age13_population_ratio  age14_population_ratio  \
0                 0.04482                 0.04167                 0.03767   
1                 0.04894                 0.04544                 0.04145   
2                 0.04765                 0.05104                 0.05273   
3                 0.04771                 0.04712                 0.04141   
4                 0.04111                 0.03967                 0.03757   

   age15_population_ratio  age16_population_ratio  age17_population_ratio  \
0                 0.03324                 0.02825                 0.01843   
1                 0.03430                 0.02730                 0.01660   
2                 0.04633                 0.03876                 0.02456   
3                 0.03502                 0.02911                 0.01942   
4                 0.02881                 0.02294                 0.01483   

   age18_population_ratio  
0                 0.01319  
1                 0.01258  
2                 0.01839  
3                 0.01368  
4                 0.01145  

[5 rows x 57 columns]

In [10]:
county_demographics_df.describe()

year          fips    population  w_population  b_population  \
count  97287.000000  97287.000000  9.728700e+04  9.728700e+04  9.728700e+04   
mean    2005.009436  30412.859950  9.365957e+04  7.529841e+04  1.252406e+04   
std        8.942825  15147.842995  3.034699e+05  2.259523e+05  5.716518e+04   
min     1990.000000   1001.000000  5.500000e+01  2.900000e+01  0.000000e+00   
25%     1997.000000  18183.000000  1.094800e+04  9.225500e+03  9.500000e+01   
50%     2005.000000  29179.000000  2.490200e+04  2.166100e+04  8.000000e+02   
75%     2013.000000  45083.000000  6.347900e+04  5.624700e+04  5.440000e+03   
max     2020.000000  99999.000000  1.009486e+07  7.301940e+06  1.437609e+06   

        o_population  nh_population  hi_population  na_population  \
count   97287.000000   9.728700e+04   9.728700e+04        97287.0   
mean     1129.919044   8.009754e+04   1.356203e+04            0.0   
std      4815.437070   2.171562e+05   1.058393e+05            0.0   
min         0.000000   4.900000e+01   0.000000e+00            0.0   
25%        40.000000   1.011600e+04   1.710000e+02            0.0   
50%       136.000000   2.327000e+04   6.350000e+02            0.0   
75%       550.000000   5.989000e+04   3.148000e+03            0.0   
max    172454.000000   5.508975e+06   4.868179e+06            0.0   

       male_population  ...  age9_population_ratio  age10_population_ratio  \
count     9.728700e+04  ...           97287.000000            97287.000000   
mean      4.598061e+04  ...               0.068098                0.067555   
std       1.489240e+05  ...               0.011292                0.010417   
min       2.900000e+01  ...               0.008400                0.013420   
25%       5.445000e+03  ...               0.059970                0.060400   
50%       1.231900e+04  ...               0.068510                0.067780   
75%       3.135050e+04  ...               0.075770                0.074500   
max       4.975410e+06  ...               0.160000                0.155960   

       age11_population_ratio  age12_population_ratio  age13_population_ratio  \
count            97287.000000            97287.000000            97287.000000   
mean                 0.064839                0.060595                0.054948   
std                  0.011904                0.014153                0.014746   
min                  0.009150                0.007710                0.005320   
25%                  0.055950                0.048960                0.044160   
50%                  0.065460                0.060580                0.052770   
75%                  0.073120                0.071100                0.064605   
max                  0.153010                0.258060                0.181820   

       age14_population_ratio  age15_population_ratio  age16_population_ratio  \
count            97287.000000            97287.000000            97287.000000   
mean                 0.048050                0.039687                0.030821   
std                  0.013994                0.011871                0.009476   
min                  0.004840                0.002570                0.000000   
25%                  0.038620                0.032010                0.024680   
50%                  0.046150                0.038470                0.029950   
75%                  0.055630                0.045940                0.035970   
max                  0.183910                0.183580                0.143400   

       age17_population_ratio  age18_population_ratio  
count            97287.000000            97287.000000  
mean                 0.021798                0.020099  
std                  0.007503                0.008808  
min                  0.000000                0.000000  
25%                  0.016890                0.014210  
50%                  0.021000                0.018590  
75%                  0.025670                0.024350  
max                  0.144440                0.195400  

[8 rows x 57 

## 1.2 Loading Fips Code DataSet




In [59]:
from google.colab import files
uploaded = files.upload()
fips_df = pd.read_csv('state_and_county_fips_master.csv')
fips_df.head(10)


Saving state_and_county_fips_master.csv to state_and_county_fips_master (6).csv


fips            name state
0     0   UNITED STATES   NaN
1  1000         ALABAMA   NaN
2  1001  Autauga County    AL
3  1003  Baldwin County    AL
4  1005  Barbour County    AL
5  1007     Bibb County    AL
6  1009   Blount County    AL
7  1011  Bullock County    AL
8  1013   Butler County    AL
9  1015  Calhoun County    AL

## 1.3 Merging the county_demographics_df with fips_df

In [192]:
county_demographics_merged_df = county_demographics_df.merge(
    fips_df,
    on='fips',
    how='left'
)
county_demographics_merged_df.head()

year  fips  population  w_population  b_population  o_population  \
0  1990  1025       27289         15579         11643            35   
1  1990  1031       40293         32869          6950           160   
2  1990  1041       13598         10068          3516            11   
3  1990  1053       35526         24377         10050          1045   
4  1990  1101      209537        119702         87856           415   

   nh_population  hi_population  na_population  male_population  ...  \
0          27196             93              0            13052  ...   
1          39831            462              0            19673  ...   
2          13576             22              0             6421  ...   
3          35378            148              0            17454  ...   
4         207933           1604              0            98854  ...   

   age11_population_ratio  age12_population_ratio  age13_population_ratio  \
0                 0.04998                 0.04482                 0.04167   
1                 0.05180                 0.04894                 0.04544   
2                 0.04574                 0.04765                 0.05104   
3                 0.05174                 0.04771                 0.04712   
4                 0.04392                 0.04111                 0.03967   

   age14_population_ratio  age15_population_ratio  age16_population_ratio  \
0                 0.03767                 0.03324                 0.02825   
1                 0.04145                 0.03430                 0.02730   
2                 0.05273                 0.04633                 0.03876   
3                 0.04141                 0.03502                 0.02911   
4                 0.03757                 0.02881                 0.02294   

   age17_population_ratio  age18_population_ratio               name  state  
0                 0.01843                 0.01319      Clarke County     AL  
1                 0.01660                 0.01258      Coffee County     AL  
2                 0.02456                 0.01839    Crenshaw County     AL  
3                 0.01942                 0.01368    Escambia County     AL  
4                 0.01483                 0.01145  Montgomery County     AL  

[5 rows x 59 columns]

### Sanity Check After Merging

In [ ]:
print(county_demographics_merged_df.columns)         # See all columns
print(county_demographics_merged_df[['fips', 'name']].head())  # Confirm correct matches
print(county_demographics_merged_df.isna().sum())    # Check for unmatched FIPS

Index(['year', 'fips', 'population', 'w_population', 'b_population',
       'o_population', 'nh_population', 'hi_population', 'na_population',
       'male_population', 'female_population', 'age0_population',
       'age1_population', 'age2_population', 'age3_population',
       'age4_population', 'age5_population', 'age6_population',
       'age7_population', 'age8_population', 'age9_population',
       'age10_population', 'age11_population', 'age12_population',
       'age13_population', 'age14_population', 'age15_population',
       'age16_population', 'age17_population', 'age18_population',
       'w_population_ratio', 'b_population_ratio', 'o_population_ratio',
       'nh_population_ratio', 'hi_population_ratio', 'na_population_ratio',
       'male_population_ratio', 'female_population_ratio',
       'age0_population_ratio', 'age1_population_ratio',
       'age2_population_ratio', 'age3_population_ratio',
       'age4_population_ratio', 'age5_population_ratio',
       'age6_popula

## 1.4 Cleaning the Merged Data

Checking for unmatched FIPS — that output tells us 174 rows in the county_demographics_df didn't find a match in fips_df, because name, state, and any other info from fips_df came back as NaN.

In [62]:
unmatched = county_demographics_merged_df[county_demographics_merged_df['name'].isna()]
print(unmatched['fips'].unique()[:10])

[ 2910  4910 51917  8911  2201  8912  8913  2280  2232  2010]


In [63]:
county_demographics_merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97287 entries, 0 to 97286
Data columns (total 59 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   year                     97287 non-null  int64  
 1   fips                     97287 non-null  int64  
 2   population               97287 non-null  int64  
 3   w_population             97287 non-null  int64  
 4   b_population             97287 non-null  int64  
 5   o_population             97287 non-null  int64  
 6   nh_population            97287 non-null  int64  
 7   hi_population            97287 non-null  int64  
 8   na_population            97287 non-null  int64  
 9   male_population          97287 non-null  int64  
 10  female_population        97287 non-null  int64  
 11  age0_population          97287 non-null  int64  
 12  age1_population          97287 non-null  int64  
 13  age2_population          97287 non-null  int64  
 14  age3_population       

Drop the unmatched rows.

In [64]:
county_demographics_merged_df = county_demographics_merged_df[county_demographics_merged_df['name'].notna()].reset_index(drop=True)
county_demographics_merged_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97113 entries, 0 to 97112
Data columns (total 59 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   year                     97113 non-null  int64  
 1   fips                     97113 non-null  int64  
 2   population               97113 non-null  int64  
 3   w_population             97113 non-null  int64  
 4   b_population             97113 non-null  int64  
 5   o_population             97113 non-null  int64  
 6   nh_population            97113 non-null  int64  
 7   hi_population            97113 non-null  int64  
 8   na_population            97113 non-null  int64  
 9   male_population          97113 non-null  int64  
 10  female_population        97113 non-null  int64  
 11  age0_population          97113 non-null  int64  
 12  age1_population          97113 non-null  int64  
 13  age2_population          97113 non-null  int64  
 14  age3_population       

In [65]:
county_demographics_merged_df = county_demographics_merged_df.rename(columns={'name': 'county'})
county_demographics_merged_df = county_demographics_merged_df.dropna()
county_demographics_merged_df = county_demographics_merged_df.drop_duplicates()
county_demographics_merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97113 entries, 0 to 97112
Data columns (total 59 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   year                     97113 non-null  int64  
 1   fips                     97113 non-null  int64  
 2   population               97113 non-null  int64  
 3   w_population             97113 non-null  int64  
 4   b_population             97113 non-null  int64  
 5   o_population             97113 non-null  int64  
 6   nh_population            97113 non-null  int64  
 7   hi_population            97113 non-null  int64  
 8   na_population            97113 non-null  int64  
 9   male_population          97113 non-null  int64  
 10  female_population        97113 non-null  int64  
 11  age0_population          97113 non-null  int64  
 12  age1_population          97113 non-null  int64  
 13  age2_population          97113 non-null  int64  
 14  age3_population       

In [66]:
county_demographics_merged_df.head()

year  fips  population  w_population  b_population  o_population  \
0  1990  1025       27289         15579         11643            35   
1  1990  1031       40293         32869          6950           160   
2  1990  1041       13598         10068          3516            11   
3  1990  1053       35526         24377         10050          1045   
4  1990  1101      209537        119702         87856           415   

   nh_population  hi_population  na_population  male_population  ...  \
0          27196             93              0            13052  ...   
1          39831            462              0            19673  ...   
2          13576             22              0             6421  ...   
3          35378            148              0            17454  ...   
4         207933           1604              0            98854  ...   

   age11_population_ratio  age12_population_ratio  age13_population_ratio  \
0                 0.04998                 0.04482                 0.04167   
1                 0.05180                 0.04894                 0.04544   
2                 0.04574                 0.04765                 0.05104   
3                 0.05174                 0.04771                 0.04712   
4                 0.04392                 0.04111                 0.03967   

   age14_population_ratio  age15_population_ratio  age16_population_ratio  \
0                 0.03767                 0.03324                 0.02825   
1                 0.04145                 0.03430                 0.02730   
2                 0.05273                 0.04633                 0.03876   
3                 0.04141                 0.03502                 0.02911   
4                 0.03757                 0.02881                 0.02294   

   age17_population_ratio  age18_population_ratio             county  state  
0                 0.01843                 0.01319      Clarke County     AL  
1                 0.01660                 0.01258      Coffee County     AL  
2                 0.02456                 0.01839    Crenshaw County     AL  
3                 0.01942                 0.01368    Escambia County     AL  
4                 0.01483                 0.01145  Montgomery County     AL  

[5 rows x 59 columns]

# Geo Mapping Zip FIPS County MSA State

## Zip FIPS Mapping


In [81]:
%%capture
!pip instsall kaggle

In [42]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [82]:
!mkdir ~/.kaggle
!cp /content/drive/MyDrive/kaggle.json ~/.kaggle/

In [83]:
!!kaggle datasets download jonathanpilafas/zip-to-fips

['Dataset URL: https://www.kaggle.com/datasets/jonathanpilafas/zip-to-fips',
 'License(s): MIT']

In [85]:
!unzip zip-to-fips.zip

Archive:  zip-to-fips.zip
  inflating: ZIP to FIPS.csv         


In [235]:
zip_fips_df = pd.read_csv('ZIP to FIPS.csv')
zip_fips_df.info()
zip_fips_df.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57608 entries, 0 to 57607
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ZIP                      57608 non-null  int64  
 1   ZIP_STRING_LEADING_ZERO  57608 non-null  object 
 2   ZIP_STRING               57608 non-null  object 
 3   FIPS                     57564 non-null  float64
 4   FIPS_STRING              57564 non-null  object 
 5   STATE                    57579 non-null  object 
 6   CNTY_NAME                56783 non-null  object 
 7   _BATCH_ID_               57608 non-null  float64
 8   _BATCH_LAST_RUN_         57608 non-null  object 
dtypes: float64(2), int64(1), object(6)
memory usage: 4.0+ MB


ZIP ZIP_STRING_LEADING_ZERO ZIP_STRING    FIPS FIPS_STRING   STATE  \
0      0                  STRING     STRING     0.0      STRING  STRING   
1  36003                   36003      36003  1001.0        1001      AL   
2  36006                   36006      36006  1001.0        1001      AL   
3  36008                   36008      36008  1001.0        1001      AL   
4  36022                   36022      36022  1001.0        1001      AL   

  CNTY_NAME  _BATCH_ID_     _BATCH_LAST_RUN_  
0    STRING        57.0  2024-03-13 21:34:03  
1   Autauga        57.0  2024-03-13 21:34:03  
2   Autauga        57.0  2024-03-13 21:34:03  
3   Autauga        57.0  2024-03-13 21:34:03  
4   Autauga        57.0  2024-03-13 21:34:03

In [236]:
# Clean Zip as 5-digit string
zip_fips_df["ZIP"] = zip_fips_df["ZIP"].apply(lambda x: f"{int(x):05}")
zip_fips_df = zip_fips_df.rename(columns={"ZIP": "zip"})

# Drop meaningless columns
columns_to_drop = ["_BATCH_ID_", "_BATCH_LAST_RUN_","ZIP_STRING_LEADING_ZERO","ZIP_STRING"]
zip_fips_df = zip_fips_df.drop(columns=columns_to_drop)

# Convert column names to lower case
zip_fips_df.columns = zip_fips_df.columns.str.lower()

zip_fips_df.info()
zip_fips_df.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57608 entries, 0 to 57607
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   zip          57608 non-null  object 
 1   fips         57564 non-null  float64
 2   fips_string  57564 non-null  object 
 3   state        57579 non-null  object 
 4   cnty_name    56783 non-null  object 
dtypes: float64(1), object(4)
memory usage: 2.2+ MB


zip    fips fips_string   state cnty_name
0  00000     0.0      STRING  STRING    STRING
1  36003  1001.0        1001      AL   Autauga
2  36006  1001.0        1001      AL   Autauga
3  36008  1001.0        1001      AL   Autauga
4  36022  1001.0        1001      AL   Autauga

## Zip MSA Mapping

In [237]:
file_path = '/content/drive/MyDrive/fs11_gpci_by_msa-ZIP.xls'
# Read the Excel file
zip_MSA_df = pd.read_excel(file_path)
zip_MSA_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43508 entries, 0 to 43507
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ZIP CODE    43508 non-null  int64  
 1   STATE       43508 non-null  object 
 2   MSA No.     43497 non-null  float64
 3   GPCI        43497 non-null  float64
 4   GPCI.1      43497 non-null  float64
 5   GPCI.2      43497 non-null  float64
 6   County No.  43496 non-null  float64
 7   Unnamed: 7  43508 non-null  int64  
 8   MSA Name    43497 non-null  object 
 9   Unnamed: 9  6 non-null      object 
dtypes: float64(5), int64(2), object(3)
memory usage: 3.3+ MB


In [238]:
# Clean Zip as 5-digit string
zip_MSA_df["ZIP CODE"] = zip_MSA_df["ZIP CODE"].apply(lambda x: f"{int(x):05}")
zip_MSA_df = zip_MSA_df.rename(columns={"ZIP CODE": "ZIP_code"})

# Rename Unnamed:7 into zip_no as int64
zip_MSA_df = zip_MSA_df.rename(columns={"Unnamed: 7": "zip_no"})

# Convert county No. as int64 and rename field
zip_MSA_df["county_no"] = zip_MSA_df["County No."].fillna(0).astype("int64")

# Drop meaningless or repetitive columns
columns_to_drop = ["MSA No.", "GPCI", "GPCI.1", "GPCI.2", "County No.", "Unnamed: 9", "STATE"]
zip_MSA_df = zip_MSA_df.drop(columns=columns_to_drop)

# Convert column names to lower case
zip_MSA_df.columns = zip_MSA_df.columns.str.lower()

In [239]:
zip_MSA_df.head(5)

zip_code  zip_no                            msa name  county_no
0    00210     210  Boston-Cambridge-Quincy, MA-NH MSA         15
1    00211     211  Boston-Cambridge-Quincy, MA-NH MSA         15
2    00212     212  Boston-Cambridge-Quincy, MA-NH MSA         15
3    00213     213  Boston-Cambridge-Quincy, MA-NH MSA         15
4    00214     214  Boston-Cambridge-Quincy, MA-NH MSA         15

## Merge & Check Effectivenss of Merge


In [240]:
# Check the number of rows before merge
print("Rows in zip_MSA_df before merge:", len(zip_MSA_df))
print("Rows in zip_fips_df before merge:", len(zip_fips_df))

Rows in zip_MSA_df before merge: 43508
Rows in zip_fips_df before merge: 57608


In [249]:
merged_zip_fips_MSA_df = pd.merge(zip_MSA_df, zip_fips_df, left_on='zip_code', right_on='zip', how='inner')

In [250]:
# Rows from zip_MSA_df that didn't merge
unmatched_zip_MSA = merged_zip_fips_MSA_df[merged_zip_fips_MSA_df['zip'].isna()]

# Rows from zip_fips_df that didn't merge
unmatched_zip_fips = merged_zip_fips_MSA_df[merged_zip_fips_MSA_df['zip_code'].isna()]

print("Unmatched rows from zip_MSA_df:", unmatched_zip_MSA)
print("Unmatched rows from zip_fips_df:", unmatched_zip_fips)

Unmatched rows from zip_MSA_df: Empty DataFrame
Columns: [zip_code, zip_no, msa name, county_no, zip, fips, fips_string, state, cnty_name]
Index: []
Unmatched rows from zip_fips_df: Empty DataFrame
Columns: [zip_code, zip_no, msa name, county_no, zip, fips, fips_string, state, cnty_name]
Index: []


In [253]:
merged_zip_fips_MSA_df.info()

print("Rows in merged_zip_fips_MSA_df after merge :", len(merged_zip_fips_MSA_df))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57141 entries, 0 to 57140
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   zip_code     57141 non-null  object 
 1   zip_no       57141 non-null  int64  
 2   msa name     57141 non-null  object 
 3   county_no    57141 non-null  int64  
 4   zip          57141 non-null  object 
 5   fips         57122 non-null  float64
 6   fips_string  57122 non-null  object 
 7   state        57125 non-null  object 
 8   cnty_name    56455 non-null  object 
dtypes: float64(1), int64(2), object(6)
memory usage: 3.9+ MB
Rows in merged_zip_fips_MSA_df after merge : 57141


In [254]:
merged_zip_fips_MSA_df.head(10)

zip_code  zip_no                                           msa name  \
0    00501     501  New York-Northern New Jersey-Long Island, NY-N...   
1    00544     544  New York-Northern New Jersey-Long Island, NY-N...   
2    00601     601                            PR NONMETROPOLITAN AREA   
3    00601     601                            PR NONMETROPOLITAN AREA   
4    00602     602            Aguadilla-Isabela-San Sebastian, PR MSA   
5    00602     602            Aguadilla-Isabela-San Sebastian, PR MSA   
6    00603     603            Aguadilla-Isabela-San Sebastian, PR MSA   
7    00603     603            Aguadilla-Isabela-San Sebastian, PR MSA   
8    00604     604            Aguadilla-Isabela-San Sebastian, PR MSA   
9    00605     605            Aguadilla-Isabela-San Sebastian, PR MSA   

   county_no    zip     fips fips_string state  cnty_name  
0        103  00501  36103.0       36103    NY    Suffolk  
1        103  00544  36103.0       36103    NY    Suffolk  
2          1  00601  72001.0       72001    PR   Adjuntas  
3          1  00601  72113.0       72113    PR      Ponce  
4          3  00602  72003.0       72003    PR     Aguada  
5          3  00602  72005.0       72005    PR  Aguadilla  
6          5  00603  72005.0       72005    PR  Aguadilla  
7          5  00603  72071.0       72071    PR    Isabela  
8          5  00604  72005.0       72005    PR  Aguadilla  
9          5  00605  72005.0       72005    PR  Aguadilla

# 2. FEMA Disaster Declarations Data Cleaning and Wrangling

In this section, we clean and wrangle FEMA disaster declaration data to prepare it for analysis. The raw data is filtered to include only disasters from the past 10 years, joined with ZIP codes using a county-level crosswalk, and aggregated to generate a ZIP-level disaster count feature for downstream housing price modeling.

The final output table, `zip_disaster_counts`, contains the number of FEMA disaster declarations per ZIP code over the last decade. This will be used as a static risk feature in our housing price model.





##2.1 Import Required Libraries



In [193]:
import pandas as pd
import numpy as np
import datetime as dt
import matplotlib.pyplot as plt
import seaborn as sns
import os
import requests

## 2.2 Load FEMA Disaster Declarations Data

Note: We loaded the full dataset here to understand the available columns. In the next step, we’ll select only the relevant ones.

In [255]:
fema_url = "https://www.fema.gov/api/open/v2/DisasterDeclarationsSummaries.csv"
fema_file_path = "DisasterDeclarationsSummaries.csv"

# Download the file if it doesn't already exist
if not os.path.exists(fema_file_path):
    response = requests.get(fema_url)
    with open(fema_file_path, "wb") as file:
        file.write(response.content)

# Load the entire dataset first (optional: you can limit rows for quick testing)
fema_raw_df = pd.read_csv(fema_file_path)

# Preview full column names
fema_raw_df.info()
fema_raw_df.head(5)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67901 entries, 0 to 67900
Data columns (total 28 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   femaDeclarationString     67901 non-null  object
 1   disasterNumber            67901 non-null  int64 
 2   state                     67901 non-null  object
 3   declarationType           67901 non-null  object
 4   declarationDate           67901 non-null  object
 5   fyDeclared                67901 non-null  int64 
 6   incidentType              67901 non-null  object
 7   declarationTitle          67901 non-null  object
 8   ihProgramDeclared         67901 non-null  int64 
 9   iaProgramDeclared         67901 non-null  int64 
 10  paProgramDeclared         67901 non-null  int64 
 11  hmProgramDeclared         67901 non-null  int64 
 12  incidentBeginDate         67901 non-null  object
 13  incidentEndDate           67092 non-null  object
 14  disasterCloseoutDate  

<ipython-input-255-774d891ceba5>:11: DtypeWarning: Columns (21) have mixed types. Specify dtype option on import or set low_memory=False.
  fema_raw_df = pd.read_csv(fema_file_path)


femaDeclarationString  disasterNumber state declarationType  \
0            EM-3611-VI            3611    VI              EM   
1            FM-5529-OR            5529    OR              FM   
2            FM-5528-OR            5528    OR              FM   
3            FM-5527-OR            5527    OR              FM   
4            EM-3611-VI            3611    VI              EM   

            declarationDate  fyDeclared    incidentType  \
0  2024-08-18T00:00:00.000Z        2024  Tropical Storm   
1  2024-08-09T00:00:00.000Z        2024            Fire   
2  2024-08-06T00:00:00.000Z        2024            Fire   
3  2024-08-02T00:00:00.000Z        2024            Fire   
4  2024-08-18T00:00:00.000Z        2024  Tropical Storm   

         declarationTitle  ihProgramDeclared  iaProgramDeclared  ...  \
0  TROPICAL STORM ERNESTO                  0                  0  ...   
1          LEE FALLS FIRE                  0                  0  ...   
2           ELK LANE FIRE                  0                  0  ...   
3    MILE MARKER 132 FIRE                  0                  0  ...   
4  TROPICAL STORM ERNESTO                  0                  0  ...   

   placeCode                          designatedArea declarationRequestNumber  \
0      99010  St. Croix (Island) (County-equivalent)                    24126   
1      99067                     Washington (County)                    24122   
2      99031                      Jefferson (County)                    24116   
3      99017                      Deschutes (County)                    24111   
4      99020   St. John (Island) (County-equivalent)                    24126   

  lastIAFilingDate  incidentId  region  designatedIncidentTypes  \
0              NaN  2024080901       2                  4,M,W,Z   
1              NaN  2024081001      10                        R   
2              NaN  2024080701      10                        R   
3              NaN  2024080301      10                        R   
4              NaN  2024080901       2                  4,M,W,Z   

                lastRefresh                                      hash  \
0  2025-03-25T22:41:27.147Z  4a8baf2d2f9117b4381e823fd96a41ea5f5c72d3   
1  2024-08-27T18:22:14.800Z  ae87cf3c6ed795015b714af7166c7c295b2b67c7   
2  2024-08-27T18:22:14.800Z  432cf0995c47e3895cea696ede5621b810460501   
3  2024-08-27T18:22:14.800Z  2f21d90cb6bc64b0d4121aa3f18d852bbb4b11fa   
4  2025-03-25T22:41:27.147Z  049f9b3d480604687cd32784486d584b44381ff4   

                                     id  
0  488c7bb4-5f20-4108-b499-99ef33953617  
1  09e3f81a-5e16-4b72-b317-1c64e0cfa59c  
2  59983f89-30bf-4888-b21b-62e8d57d9aac  
3  8d13ecf0-bc2f-496b-8c9f-b2e73da832a0  
4  4576151c-0d18-478f-a342-1f89463a736e  

[5 rows x 28 columns]

## 2.3 Select Relevant Columns：

In [265]:
# Select only the columns we need
fema_df = fema_raw_df[[
    'disasterNumber', 'state', 'designatedArea', 'declarationDate', 'fyDeclared', 'incidentType',
    'designatedArea',
    'fipsStateCode', 'fipsCountyCode'
]].copy()

# Convert to datetime
fema_df['declarationDate'] = pd.to_datetime(fema_df['declarationDate'])

# Construct full 5-digit county-level FIPS code
fema_df['fips'] = (
    fema_df['fipsStateCode'].astype(str).str.zfill(2) +
    fema_df['fipsCountyCode'].astype(str).str.zfill(3)
)

# Preview cleaned dataframe
fema_df.info()
fema_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67901 entries, 0 to 67900
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype              
---  ------           --------------  -----              
 0   disasterNumber   67901 non-null  int64              
 1   state            67901 non-null  object             
 2   designatedArea   67901 non-null  object             
 3   declarationDate  67901 non-null  datetime64[ns, UTC]
 4   fyDeclared       67901 non-null  int64              
 5   incidentType     67901 non-null  object             
 6   designatedArea   67901 non-null  object             
 7   fipsStateCode    67901 non-null  int64              
 8   fipsCountyCode   67901 non-null  int64              
 9   fips             67901 non-null  object             
dtypes: datetime64[ns, UTC](1), int64(4), object(5)
memory usage: 5.2+ MB


disasterNumber state                          designatedArea  \
0            3611    VI  St. Croix (Island) (County-equivalent)   
1            5529    OR                     Washington (County)   
2            5528    OR                      Jefferson (County)   
3            5527    OR                      Deschutes (County)   
4            3611    VI   St. John (Island) (County-equivalent)   

            declarationDate  fyDeclared    incidentType  \
0 2024-08-18 00:00:00+00:00        2024  Tropical Storm   
1 2024-08-09 00:00:00+00:00        2024            Fire   
2 2024-08-06 00:00:00+00:00        2024            Fire   
3 2024-08-02 00:00:00+00:00        2024            Fire   
4 2024-08-18 00:00:00+00:00        2024  Tropical Storm   

                           designatedArea  fipsStateCode  fipsCountyCode  \
0  St. Croix (Island) (County-equivalent)             78              10   
1                     Washington (County)             41              67   
2                      Jefferson (County)             41              31   
3                      Deschutes (County)             41              17   
4   St. John (Island) (County-equivalent)             78              20   

    fips  
0  78010  
1  41067  
2  41031  
3  41017  
4  78020

## 2.4 Filter Data from the Last 10 Years

In [266]:
from pandas.tseries.offsets import DateOffset

# Today's date and dynamic 10-year cutoff
today = pd.Timestamp.now(tz='UTC')
cutoff_date = today - DateOffset(years=10)

# Filter FEMA data
fema_recent_df = fema_df[fema_df['declarationDate'] >= cutoff_date].copy()



# FEMA History from 1990 to 2020 (Match Demographic df timeline)

# Filter rows where 'fyDeclared' is between 1990 and 2020 (inclusive)
filtered_fema_df = fema_df[(fema_df['fyDeclared'] >= 1990) & (fema_recent_df['fyDeclared'] <= 2020)]



# Check the date range

# print("Earliest disaster:", fema_recent_df['declarationDate'].min())
# print("Latest disaster:", fema_recent_df['declarationDate'].max())
# print("Total records:", fema_recent_df.shape[0])

print("Earliest disaster:", filtered_fema_df['declarationDate'].min())
print("Latest disaster:", filtered_fema_df['declarationDate'].max())
print("Total records:", filtered_fema_df.shape[0])


# Preview
filtered_fema_df.head()


Earliest disaster: 2015-04-08 00:00:00+00:00
Latest disaster: 2020-09-28 00:00:00+00:00
Total records: 15913


disasterNumber state                            designatedArea  \
6             4312    CA  Resighini Rancheria (Indian Reservation)   
7             4251    AL                          Autauga (County)   
8             4251    AL                          Barbour (County)   
9             4251    AL                           Blount (County)   
14            4251    AL                          Bullock (County)   

             declarationDate  fyDeclared  incidentType  \
6  2017-05-02 00:00:00+00:00        2017  Severe Storm   
7  2016-01-21 00:00:00+00:00        2016  Severe Storm   
8  2016-01-21 00:00:00+00:00        2016  Severe Storm   
9  2016-01-21 00:00:00+00:00        2016  Severe Storm   
14 2016-01-21 00:00:00+00:00        2016  Severe Storm   

                              designatedArea  fipsStateCode  fipsCountyCode  \
6   Resighini Rancheria (Indian Reservation)              6               0   
7                           Autauga (County)              1               1   
8                           Barbour (County)              1               5   
9                            Blount (County)              1               9   
14                          Bullock (County)              1              11   

     fips  
6   06000  
7   01001  
8   01005  
9   01009  
14  01011

## 2.5 Clean FEMA Data (Standardize & Prepare for Join) --->> MOVED TO 2.3

## 2.6 Load County-to-ZIP Mapping Table from Kaggle

In [212]:
# Install kagglehub if not already installed
!pip install kagglehub --quiet

import kagglehub
import os
import pandas as pd

# Download dataset from Kaggle
path = kagglehub.dataset_download("danofer/zipcodes-county-fips-crosswalk")

print("Path to dataset files:", path)

# List files
files = os.listdir(path)
print("Files in dataset:", files)

# Load the crosswalk CSV
file_path = os.path.join(path, "ZIP-COUNTY-FIPS_2017-06.csv")
zip_crosswalk_df = pd.read_csv(file_path)

zip_crosswalk_df['fips'] = zip_crosswalk_df['STCOUNTYFP'].astype(str).str.zfill(5)


# Preview
zip_crosswalk_df.head()


Path to dataset files: /root/.cache/kagglehub/datasets/danofer/zipcodes-county-fips-crosswalk/versions/1
Files in dataset: ['ZIP-COUNTY-FIPS_2017-06.csv']


ZIP      COUNTYNAME STATE  STCOUNTYFP CLASSFP   fips
0  36003  Autauga County    AL        1001      H1  01001
1  36006  Autauga County    AL        1001      H1  01001
2  36067  Autauga County    AL        1001      H1  01001
3  36066  Autauga County    AL        1001      H1  01001
4  36703  Autauga County    AL        1001      H1  01001


## 2.7 Merge FEMA with ZIP (via FIPS)

In [267]:
fema_with_zip = fema_recent_df.merge(
    zip_crosswalk_df[['ZIP', 'fips']],
    on='fips',
    how='left'
)

# Drop FEMA rows with no ZIP match
#fema_with_zip = fema_with_zip.dropna(subset=['ZIP'])

# Preveiw
fema_with_zip.head()


disasterNumber state                          designatedArea  \
0            3611    VI  St. Croix (Island) (County-equivalent)   
1            3611    VI  St. Croix (Island) (County-equivalent)   
2            3611    VI  St. Croix (Island) (County-equivalent)   
3            5529    OR                     Washington (County)   
4            5529    OR                     Washington (County)   

            declarationDate  fyDeclared    incidentType  \
0 2024-08-18 00:00:00+00:00        2024  Tropical Storm   
1 2024-08-18 00:00:00+00:00        2024  Tropical Storm   
2 2024-08-18 00:00:00+00:00        2024  Tropical Storm   
3 2024-08-09 00:00:00+00:00        2024            Fire   
4 2024-08-09 00:00:00+00:00        2024            Fire   

                           designatedArea  fipsStateCode  fipsCountyCode  \
0  St. Croix (Island) (County-equivalent)             78              10   
1  St. Croix (Island) (County-equivalent)             78              10   
2  St. Croix (Island) (County-equivalent)             78              10   
3                     Washington (County)             41              67   
4                     Washington (County)             41              67   

    fips      ZIP  
0  78010    850.0  
1  78010    840.0  
2  78010    820.0  
3  41067  97078.0  
4  41067  97133.0

In [215]:
print(f"Total FEMA records after 10-year filter: {fema_recent_df.shape[0]}")
print(f"FEMA records after ZIP merge: {fema_with_zip.shape[0]}")
print(f"Number of FEMA records with ZIP: {fema_with_zip['ZIP'].notna().sum()}")
print(f"Number of FEMA records missing ZIP: {fema_with_zip['ZIP'].isna().sum()}")

Total FEMA records after 10-year filter: 23427
FEMA records after ZIP merge: 389086
Number of FEMA records with ZIP: 388152
Number of FEMA records missing ZIP: 934


## 2.6-7: Merge #2 FEMA with Zip and MSA from 1990-2020 (JY)

In this section, we validate whether FEMA disaster records were successfully joined with ZIP codes using fips codes. We check the match rate, ZIP assignment distribution, and inspect records with missing ZIPs.

In [271]:
fema_with_zip_MSA_df = pd.merge(filtered_fema_df, merged_zip_fips_MSA_df, left_on='fips', right_on='fips_string', how='inner')

# Preveiw
fema_with_zip_MSA_df.head()

disasterNumber state_x   designatedArea           declarationDate  \
0            5375      WY  Albany (County) 2020-09-26 00:00:00+00:00   
1            5375      WY  Albany (County) 2020-09-26 00:00:00+00:00   
2            5375      WY  Albany (County) 2020-09-26 00:00:00+00:00   
3            5375      WY  Albany (County) 2020-09-26 00:00:00+00:00   
4            5375      WY  Albany (County) 2020-09-26 00:00:00+00:00   

   fyDeclared incidentType   designatedArea  fipsStateCode  fipsCountyCode  \
0        2020         Fire  Albany (County)             56               1   
1        2020         Fire  Albany (County)             56               1   
2        2020         Fire  Albany (County)             56               1   
3        2020         Fire  Albany (County)             56               1   
4        2020         Fire  Albany (County)             56               1   

  fips_x zip_code  zip_no                 msa name  county_no    zip   fips_y  \
0  56001    82051   82051  WY NONMETROPOLITAN AREA          1  82051  56001.0   
1  56001    82052   82052  WY NONMETROPOLITAN AREA          1  82052  56001.0   
2  56001    82055   82055  WY NONMETROPOLITAN AREA          1  82055  56001.0   
3  56001    82058   82058  WY NONMETROPOLITAN AREA          1  82058  56001.0   
4  56001    82063   82063  WY NONMETROPOLITAN AREA          1  82063  56001.0   

  fips_string state_y cnty_name  
0       56001      WY    Albany  
1       56001      WY    Albany  
2       56001      WY    Albany  
3       56001      WY    Albany  
4       56001      WY    Albany

In [273]:
print(f"Total FEMA records 1990-2020 filter: {filtered_fema_df.shape[0]}")
print(f"FEMA records after ZIP merge: {fema_with_zip_MSA_df.shape[0]}")
print(f"Number of FEMA records with ZIP: {fema_with_zip_MSA_df['zip_code'].notna().sum()}")
print(f"Number of FEMA records missing ZIP: {fema_with_zip_MSA_df['zip_code'].isna().sum()}")

Total FEMA records 1990-2020 filter: 15913
FEMA records after ZIP merge: 225306
Number of FEMA records with ZIP: 225306
Number of FEMA records missing ZIP: 0


## 2.8: Merge Validation — FEMA and ZIP Mapping

In this section, we validate whether FEMA disaster records were successfully joined with ZIP codes using fips codes. We check the match rate, ZIP assignment distribution, and inspect records with missing ZIPs.


### 2.8.1 Match Rate Check
This checks whether the merge was successful for a significant portion of the data.

In [216]:
total_records = fema_with_zip.shape[0]
matched_zip_count = fema_with_zip['ZIP'].notna().sum()
missing_zip_count = fema_with_zip['ZIP'].isna().sum()
match_rate = matched_zip_count / total_records

print(f"Total FEMA records (10 yrs):     {total_records}")
print(f"Matched ZIPs:                    {matched_zip_count}")
print(f"Missing ZIPs:                    {missing_zip_count}")
print(f"ZIP Match Rate:                  {match_rate:.2%}")


Total FEMA records (10 yrs):     389086
Matched ZIPs:                    388152
Missing ZIPs:                    934
ZIP Match Rate:                  99.76%


### 2.8.2 ZIP Duplication per FIPS

One county (FIPS) may correspond to multiple ZIP codes. We inspect the distribution to ensure the ZIP mapping is reasonable.

In [217]:
zip_per_fips = (
    fema_with_zip[['fips', 'ZIP']]
    .drop_duplicates()
    .groupby('fips')
    .size()
    .reset_index(name='zip_count')
)

zip_per_fips['zip_count'].describe()

count    3277.000000
mean       16.155935
std        18.770522
min         1.000000
25%         7.000000
50%        12.000000
75%        18.000000
max       494.000000
Name: zip_count, dtype: float64

###2.8.3 Missing ZIPs (Rows That Didn't Match)
Identify FEMA records from counties not covered in the ZIP mapping dataset.

In [218]:
missing_zip_df = fema_with_zip[fema_with_zip['ZIP'].isna()]
print(f"Number of FEMA records without ZIP after merge: {missing_zip_df.shape[0]}")
missing_zip_df[['state', 'designatedArea', 'fips']].drop_duplicates().head(10)

Number of FEMA records without ZIP after merge: 934


state                            designatedArea  \
58       CA  Resighini Rancheria (Indian Reservation)   
123      ID              Nez Perce Indian Reservation   
165      WA               Colville Indian Reservation   
198      WA                        Yakama Reservation   
452      NM                           Mescalero Tribe   
3083     MT                   Crow Indian Reservation   
5627     OR           Warm Springs Indian Reservation   
6452     AZ             San Carlos Indian Reservation   
7091     CA                Morongo Indian Reservation   
11913    CA    Hopland Rancheria (Indian Reservation)   

                                 designatedArea   fips  
58     Resighini Rancheria (Indian Reservation)  06000  
123                Nez Perce Indian Reservation  16000  
165                 Colville Indian Reservation  53000  
198                          Yakama Reservation  53000  
452                             Mescalero Tribe  35000  
3083                    Crow Indian Reservation  30000  
5627            Warm Springs Indian Reservation  41000  
6452              San Carlos Indian Reservation  04000  
7091                 Morongo Indian Reservation  06000  
11913    Hopland Rancheria (Indian Reservation)  06000

#### 2.8.3 ZIP Missing Analysis

After merging FEMA disaster data with ZIP codes using FIPS codes, **934** records could not be matched.

Top unmatched designatedArea examples:
*   Resighini Rancheria (Indian Reservation), CA
*   Nez Perce Indian Reservation, ID
*   Yakama Reservation, WA
*   Warm Springs Indian Reservation, OR
*   Crow Indian Reservation, MT
*   San Carlos Indian Reservation, AZ
*   Hopland Rancheria (Indian Reservation), CA



**Possible reasons:**
- Indian Reservations and tribal lands often don't align with standard county or ZIP boundaries

- Non-standard area names that don’t match ZIP datasets (e.g., Rancheria, Reservation)

**Action:** These records can be excluded

In [219]:
# Drop FEMA records with missing ZIPs
fema_with_zip['zip_matched'] = fema_with_zip['ZIP'].notna()
fema_matched_df = fema_with_zip[fema_with_zip['zip_matched']].copy().reset_index(drop=True)

# Preview
print(fema_matched_df.shape)
fema_matched_df.head()

(388152, 12)


disasterNumber state                          designatedArea  \
0            3611    VI  St. Croix (Island) (County-equivalent)   
1            3611    VI  St. Croix (Island) (County-equivalent)   
2            3611    VI  St. Croix (Island) (County-equivalent)   
3            5529    OR                     Washington (County)   
4            5529    OR                     Washington (County)   

            declarationDate  fyDeclared    incidentType  \
0 2024-08-18 00:00:00+00:00        2024  Tropical Storm   
1 2024-08-18 00:00:00+00:00        2024  Tropical Storm   
2 2024-08-18 00:00:00+00:00        2024  Tropical Storm   
3 2024-08-09 00:00:00+00:00        2024            Fire   
4 2024-08-09 00:00:00+00:00        2024            Fire   

                           designatedArea  fipsStateCode  fipsCountyCode  \
0  St. Croix (Island) (County-equivalent)             78              10   
1  St. Croix (Island) (County-equivalent)             78              10   
2  St. Croix (Island) (County-equivalent)             78              10   
3                     Washington (County)             41              67   
4                     Washington (County)             41              67   

    fips      ZIP  zip_matched  
0  78010    850.0         True  
1  78010    840.0         True  
2  78010    820.0         True  
3  41067  97078.0         True  
4  41067  97133.0         True

### 2.9.4 FEMA-ZIP Merge Summary

<!-- We compare the FEMA-ZIP dataset before and after filtering unmatched records:

| Metric                                  | Before Filtering (`fema_with_zip`) | After Filtering (`fema_matched_df`) |
|-----------------------------------------|-------------------------------------|--------------------------------------|
| Total rows                              | 390,248                             | 387,209                              |
| Records with matched ZIP                | 387,209                             | 387,209                              |
| Records with missing ZIP                | 3,039                               | 0                                    |
| Unique disaster-county combinations     | 24,904                              | 21,888                               |
| Match rate (disaster-county level)      | 87.89%                              | 100%                                 |
| Unique ZIP codes                        | 38,377                              | 38,377                               |

**Optimization Summary:**
- Removed 3,039 unmatched records (0.8%) from U.S. territories and special cases
- Ensured 100% ZIP code coverage for all retained FEMA disaster-county entries
- Final dataset `fema_matched_df` is clean, ZIP-aligned, and ready for ZIP-level aggregation -->


In [220]:
# FEMA-ZIP Merge Summary (auto-generated)

# Total rows
rows_before = fema_with_zip.shape[0]
rows_after = fema_matched_df.shape[0]

# ZIP matching stats
matched_rows = fema_with_zip['ZIP'].notna().sum()
missing_rows = fema_with_zip['ZIP'].isna().sum()
match_rate = matched_rows / rows_before
final_match_rate = 1.0  # because all rows in fema_matched_df have ZIPs

# ZIPs per FIPS (before filtering)
zip_per_fips = (
    fema_with_zip[['fips', 'ZIP']]
    .drop_duplicates()
    .groupby('fips')
    .size()
    .reset_index(name='zip_count')
)

# Summary stats
print("FEMA-ZIP Merge Summary:")
print(f"Total FEMA records (before filtering):   {rows_before}")
print(f"Total FEMA records (after filtering):    {rows_after}")
print(f"Matched ZIPs:                             {matched_rows}")
print(f"Missing ZIPs:                             {missing_rows}")
print(f"ZIP Match Rate:                           {match_rate:.2%}")
print(f"ZIP Match Rate (final dataset):           {final_match_rate:.2%}")
print(f"Unique county FIPS:                       {zip_per_fips.shape[0]}")
print(f"Median ZIPs per county:                   {zip_per_fips['zip_count'].median():.0f}")
print(f"Max ZIPs per county:                      {zip_per_fips['zip_count'].max()}")


FEMA-ZIP Merge Summary:
Total FEMA records (before filtering):   389086
Total FEMA records (after filtering):    388152
Matched ZIPs:                             388152
Missing ZIPs:                             934
ZIP Match Rate:                           99.76%
ZIP Match Rate (final dataset):           100.00%
Unique county FIPS:                       3277
Median ZIPs per county:                   12
Max ZIPs per county:                      494


##2.10 Disaster Count Aggregation by ZIP (Last 10 Years)

To support downstream analysis, we compute the number of FEMA disaster declarations in the past 10 years for each ZIP code. This gives a static risk exposure feature per ZIP that can be merged with housing data.

The aggregation is based on the cleaned FEMA dataset `fema_matched_df`, which includes only matched ZIP-level records between 2013 and 2025.

In [221]:
# Count the number of disasters per ZIP code
zip_disaster_counts = (
    fema_matched_df.groupby('ZIP')['disasterNumber']
    .nunique()
    .reset_index(name='disaster_count_10yrs')
)

# Format ZIPs as 5-digit strings
# zip_disaster_counts['ZIP'] = zip_disaster_counts['ZIP'].astype(int).astype(str).str.zfill(5)

# Make sure ZIP is stored as integer (drop decimals)
zip_disaster_counts['ZIP'] = zip_disaster_counts['ZIP'].astype(int)

# Reset index for cleanliness
zip_disaster_counts = zip_disaster_counts.reset_index(drop=True)

# Preview top ZIPs by disaster count
zip_disaster_counts.sort_values(by='disaster_count_10yrs', ascending=False).head(10)


ZIP  disaster_count_10yrs
35885  91710                    41
35884  91709                    41
35934  91786                    41
35920  91766                    41
35623  90623                    39
35625  90630                    39
35626  90631                    39
35630  90638                    39
35680  90808                    39
36711  93560                    38

## Summary of Disaster Count Aggregation

The resulting table `zip_disaster_counts` contains the number of unique FEMA disaster declarations per ZIP code over the past 10 years (2013–2025). Each row represents a ZIP and its corresponding disaster exposure:

- **ZIP**: 5-digit ZIP code (int)
- **disaster_count_10yrs**: total number of FEMA disaster declarations in that ZIP from 2013 onward

This table will serve as a **static risk feature** that can be joined with the main housing dataset using the ZIP code as key. It helps capture regional environmental vulnerability, which may influence housing prices.


# HPI Dataframe by Zip

## Uploading Data

In [316]:
file_path = '/content/drive/MyDrive/hpi_at_bdl_zip5.csv'
# Read the file
hpi_df = pd.read_csv(file_path)
hpi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 668175 entries, 0 to 668174
Data columns (total 13 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Five-Digit ZIP Code  668175 non-null  int64  
 1   Year                 668175 non-null  int64  
 2   Annual Change (%)    626080 non-null  float64
 3   HPI                  651400 non-null  float64
 4   HPI with 1990 base   396604 non-null  float64
 5   HPI with 2000 base   573289 non-null  float64
 6   Unnamed: 6           0 non-null       float64
 7   Unnamed: 7           0 non-null       float64
 8   Unnamed: 8           0 non-null       float64
 9   Unnamed: 9           0 non-null       float64
 10  Unnamed: 10          0 non-null       float64
 11  Unnamed: 11          0 non-null       float64
 12  Unnamed: 12          0 non-null       float64
dtypes: float64(11), int64(2)
memory usage: 66.3 MB


In [317]:
# Drop meaningless columns
columns_to_drop = ["HPI with 2000 base", "Unnamed: 6","Unnamed: 7","Unnamed: 8","Unnamed: 9","Unnamed: 10","Unnamed: 11","Unnamed: 12",]
hpi_df = hpi_df.drop(columns=columns_to_drop)

# Convert "Five-Digit ZIP Code" to string and rename the column to zip_code
hpi_df["Five-Digit ZIP Code"] = hpi_df["Five-Digit ZIP Code"].apply(lambda x: f"{int(x):05}")
hpi_df = hpi_df.rename(columns={"Five-Digit ZIP Code": "zip_code"})

# Filter timeline from 1990-2020
filtered_hpi_df = hpi_df[(hpi_df['Year'] >= 1990) & (hpi_df['Year'] <= 2020)]

# Rename columns
filtered_hpi_df = filtered_hpi_df.rename(columns={"Annual Change (%)": "annual_change"})
filtered_hpi_df = filtered_hpi_df.rename(columns={"HPI with 1990 base": "hpi_1990"})


filtered_hpi_df = filtered_hpi_df.reset_index(drop=True)
filtered_hpi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 512396 entries, 0 to 512395
Data columns (total 5 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   zip_code       512396 non-null  object 
 1   Year           512396 non-null  int64  
 2   annual_change  493590 non-null  float64
 3   HPI            507056 non-null  float64
 4   hpi_1990       283174 non-null  float64
dtypes: float64(3), int64(1), object(1)
memory usage: 19.5+ MB


In [318]:
filtered_hpi_df.head(10)

zip_code  Year  annual_change     HPI  hpi_1990
0    01001  1990          -1.51  191.16    100.00
1    01001  1991          -5.50  180.65     94.50
2    01001  1992          -1.56  177.84     93.03
3    01001  1993           0.56  178.84     93.55
4    01001  1994          -3.05  173.39     90.70
5    01001  1995           1.91  176.70     92.44
6    01001  1996           0.12  176.91     92.55
7    01001  1997          -1.33  174.56     91.32
8    01001  1998           4.47  182.36     95.39
9    01001  1999           1.60  185.27     96.92

## Check Data Quality for Housing Price Time Series

In [336]:
# Check the number of rows in dataframe
print("Rows in filtered_hpi_df before dropping NA:\n", len(filtered_hpi_df))
print("\n")

# Check Nulls or missing values
missing_values = filtered_hpi_df.isna().sum()
missing_percentage = (missing_values / len(filtered_hpi_df)) * 100

print("Missing value breakdown:\n", missing_values)
print("\n")
print("Missing Percentage:\n", missing_percentage)
print("\n")
# Find rows where only 'HPI' is NaN and other columns have non-NaN values
only_hpi_na = filtered_hpi_df[filtered_hpi_df['HPI'].isna() & filtered_hpi_df.drop(columns=['HPI']).notna().all(axis=1)]
# Display how many rows have only 'HPI' NaN and not NaN in other columns
print(f"Rows with NaN in 'HPI' but not in other columns: {only_hpi_na.shape[0]}")

Rows in filtered_hpi_df before dropping NA:
 512396


Missing value breakdown:
 zip_code              0
Year                  0
annual_change     18806
HPI                5340
hpi_1990         229222
dtype: int64


Missing Percentage:
 zip_code          0.000000
Year              0.000000
annual_change     3.670208
HPI               1.042163
hpi_1990         44.735322
dtype: float64


Rows with NaN in 'HPI' but not in other columns: 0


In [341]:
# Drop hpi_1990 column since there is a large amount of missing value and won't be good reference data point
filtered_hpi_df = filtered_hpi_df.drop(columns=['hpi_1990'])
filtered_hpi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 512396 entries, 0 to 512395
Data columns (total 4 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   zip_code       512396 non-null  object 
 1   Year           512396 non-null  int64  
 2   annual_change  493590 non-null  float64
 3   HPI            507056 non-null  float64
dtypes: float64(2), int64(1), object(1)
memory usage: 15.6+ MB


In [348]:
# Drop NAs in HPI and using HPI as major housing price prediction
# // Potetial Alternative Treatment on Missing values is to get Average of HPI for ZIPs if they are in the same MSA or County

cleaned_hpi_df = filtered_hpi_df.dropna(subset=['annual_change', 'HPI'], how='all')
print("Rows in cleaned_hpi_df after dropping NA:\n", len(cleaned_hpi_df))

cleaned_hpi_df.head()

Rows in cleaned_hpi_df after dropping NA:
 507056


zip_code  Year  annual_change     HPI
0    01001  1990          -1.51  191.16
1    01001  1991          -5.50  180.65
2    01001  1992          -1.56  177.84
3    01001  1993           0.56  178.84
4    01001  1994          -3.05  173.39

## Transform Data for Zip code level with annual HPI info

In [351]:
# Convert the HPI table by Zip level with annual HPI info appended for each year


# Pivot the data: reshape so each year’s values become their own set of columns
pivoted_hpi_df = filtered_hpi_df.pivot_table(
    index='zip_code',
    columns='Year',
    values=['HPI', 'annual_change']
)

# Flatten the multi-level column index
pivoted_hpi_df.columns = [f'{metric}_{year}' for metric, year in pivoted_hpi_df.columns]

# Reset index to turn zip_code back into a column
pivoted_hpi_df = pivoted_hpi_df.reset_index()

# Display the result
print(pivoted_hpi_df.head())
print("\n")
print("Rows in pivoted_hpi_df", len(pivoted_hpi_df))

  zip_code  HPI_1990  HPI_1991  HPI_1992  HPI_1993  HPI_1994  HPI_1995  \
0    01001    191.16    180.65    177.84    178.84    173.39    176.70   
1    01002    204.49    198.94    197.17    196.69    200.77    196.13   
2    01005       NaN       NaN    100.00    100.17     94.10    100.28   
3    01007    160.18    155.68    151.99    153.49    151.84    152.76   
4    01008       NaN       NaN       NaN       NaN       NaN       NaN   

   HPI_1996  HPI_1997  HPI_1998  ...  annual_change_2011  annual_change_2012  \
0    176.91    174.56    182.36  ...               -2.63               -4.48   
1    199.60    209.49    213.47  ...               -1.63               -2.35   
2     99.63    101.98    105.71  ...               -8.32               -4.70   
3    154.99    156.68    160.90  ...               -2.96               -1.65   
4       NaN       NaN       NaN  ...               -6.79               -0.45   

   annual_change_2013  annual_change_2014  annual_change_2015  \
0        

In [356]:
# merged_hpi_df = pd.merge(merged_zip_county_MSA_df, pivoted_hpi_df, left_on='zip_code', right_on='zip_code', how='inner')

merged_hpi_df = pd.merge(
    merged_zip_county_MSA_df,
    pivoted_hpi_df,
    left_on='zip_code',
    right_on='zip_code',
    how='outer',  # use 'outer' to catch all unmatched rows from both sides
    indicator=True
)

In [359]:
# Rows only from left DataFrame (no match from pivoted_hpi_df)
unmatched_left = merged_hpi_df[merged_hpi_df['_merge'] == 'left_only']

# Rows only from right DataFrame (no match from merged_zip_county_MSA_df)
unmatched_right = merged_hpi_df[merged_hpi_df['_merge'] == 'right_only']

# Zip codes from left DataFrame that had no match in right
print("Unmatched zip codes from left DataFrame:")
print(unmatched_left['zip_code'].unique())

# Zip codes from right DataFrame that had no match in left
print("Unmatched zip codes from right DataFrame:")
print(unmatched_right['zip_code'].unique())
print("\n")

missing_from_right = (merged_hpi_df['_merge'] == 'left_only').sum()
missing_from_left = (merged_hpi_df['_merge'] == 'right_only').sum()
print(f"Zip codes missing from pivoted_hpi_df (right): {missing_from_right}")
print(f"Zip codes missing from merged_zip_county_MSA_df (left): {missing_from_left}")

# merged_hpi_df.drop(columns=['_merge'], inplace=True)

Unmatched zip codes from left DataFrame:
['10001' '10002' '10003' ... '99928' '99929' '99950']
Unmatched zip codes from right DataFrame:
['01001' '01002' '01005' ... '87547' '97003' '97703']


Zip codes missing from pivoted_hpi_df (right): 26955
Zip codes missing from merged_zip_county_MSA_df (left): 1825


In [355]:
print(merged_hpi_df.head())
print("\n")
print("Rows in pivoted_hpi_df", len(pivoted_hpi_df))

  zip_code  zip_no                                           msa name  \
0    10001   10001  New York-Northern New Jersey-Long Island, NY-N...   
1    10002   10002  New York-Northern New Jersey-Long Island, NY-N...   
2    10003   10003  New York-Northern New Jersey-Long Island, NY-N...   
3    10004   10004  New York-Northern New Jersey-Long Island, NY-N...   
4    10005   10005  New York-Northern New Jersey-Long Island, NY-N...   

   county_no     fips fips_string state cnty_name  HPI_1990  HPI_1991  ...  \
0         61  36061.0       36061    NY  New York       NaN       NaN  ...   
1         61  36061.0       36061    NY  New York       NaN       NaN  ...   
2         61  36061.0       36061    NY  New York       NaN       NaN  ...   
3         61  36061.0       36061    NY  New York       NaN       NaN  ...   
4         61  36061.0       36061    NY  New York       NaN       NaN  ...   

   annual_change_2011  annual_change_2012  annual_change_2013  \
0                 NaN      

## Merging HPI time series with Demographic and FEMA dataframe

In [ ]:
# Merge HPI with Demographic and FEMA dataframe

